# Computação Científica 1 - 2023.2 - Prof. Adriano Côrtes

## Teste 1

***

### Questão 1
Explique o comportamento (surpreendente?) do código abaixo

In [9]:
d = 8
e = 2
from math import *
sqrt(d ** e)

16.88210319127114

O motivo é simples, ao realizarmos a importação da biblioteca math através do seguinte trecho from math import *, estaremos importando também "e" como sendo o número de euler dado por 2.71828... e com isso ao realizarmos o cálculo sqrt(d**e) obteremos aquele valor 16.88210319127114. Com isso vemos que é importante saber o que estamos importando ao rodarmos o código acima

### Questão 2
Como sabemos números reais são representados no computador por números de precisão finita, os números em ponto-flutuante. No Python, assim como na maioria das outras linguagens de programação predomina o padrão IEEE-754, em particular a representação em precisão dupla (*double precision*) é o default em Python. Podemos investigar essas informações por meio do módulo `sys`. Essas informações estão atributo `sys.float_info`:

In [10]:
import sys
print(sys.float_info)
print('O menor número representável em ponto-flutuante de precisão dupla é {}'.format(sys.float_info.min))
print('O maior número representável em ponto-flutuante de precisão dupla é {}'.format(sys.float_info.max))

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)
O menor número representável em ponto-flutuante de precisão dupla é 2.2250738585072014e-308
O maior número representável em ponto-flutuante de precisão dupla é 1.7976931348623157e+308


1. O que acontece se você tentar avaliar a expressão $e^{1000}$ que gera um número maior que o maior número representável?
2. Como você calcularia a hipotenusa $c$ de um triângulo retângulo cujos catetos são $a = 1.5 \times 10^{200}$ e $b = 3.5 \times 10^{201}$? **Note que calcular $a^2$ e $b^2$ não é uma opção, como no item anterior**. Verifique seu resultado usando a função `math.hypot`

***

***

### Questão 3
Usando uma tupla de strings com o nome dos dígitos de $0$ a $9$, crie um trecho de código em Python que imprima até a sétima casas decimais de $\pi$ na forma de texto, como:

`três ponto um quatro um cinco nove dois seis`

***

### Questão 4
Escreva um trecho que código Python que imprima de maneira formatada (bonitinha) as oito primeiro linhas do triângulo de Pascal.

***

### Questão 5
1. Crie uma função que receba como entrada um número inteiro $N$ e retorna uma lista com os primeiros $N$ números de Fibonacci.
2. A chamada [*lei de Benford*](https://pt.wikipedia.org/wiki/Lei_de_Benford) é uma lei baseada na observação sobre a frequência do primeiro dígito $d \in \{1,2,\ldots,9\}$ de uma determinada coleção de números (*data sets*). Segundo essa lei a distribuição logaritmica, portanto não-uniforme, dada por $$P(d) = \log_{10} \left( \dfrac{d+1}{d} \right),$$ ou seja, número começando com $1$ são mais frequentes que os começando com $2$, e assim em diante, com efeito temos aproximadamente

| $d$ | $P(d)$ |
|-----|--------|
|1    | 0.301  |
|2    | 0.176  |
|3    | 0.125  |
|4    | 0.097  |
|5    | 0.079  |
|6    | 0.067  |
|7    | 0.058  |
|8    | 0.051  |
|9    | 0.046  |

Usando a função criada no item anterior calcule a distribuição dos primeiros dígitos da sequência de Fibonacci para $N=500$ e $N=1000$, e avalie se ela segue a lei de Benford.

***

### Questão 6
Escreva um loop usando `while` para calcular a [média aritmética-geometrica](https://en.wikipedia.org/wiki/Arithmetic%E2%80%93geometric_mean) (AGM) de dois números reais positivos, $x$ e $y$, definida como o limite das sequência:

$$ 
\begin{align}
a_{n+1} &= \frac{1}{2} (a_n + g_n), \\
g_{n+1} &= \sqrt{a_n g_n},
\end{align}
$$

começando com $a_0 = x$ e $g_0 = y$. Ambas as sequências convergem para o mesmo número, denotado por $\textrm{agm}(x,y)$. Use sua implementação para calcular a constante de *Gauss*, $G = 1/\textrm{agm}(1,\sqrt{2})$.

**Importante:** Você terá que usar um critério de parada para iteração. 

***